pkg/gpu/cuda/fatbin.go

```diff
@@ -0,0 +1,475 @@
+	UncompressedPayloadSize uint64
+}
+
+// ParseFatbinFromELFFilePath opens the given path and parses the resulting ELF for CUDA kernels
+func ParseFatbinFromELFFilePath(path string) (*Fatbin, error) {
+	elfFile, err := elf.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open ELF file %s: %w", path, err)
+	}
+	defer elfFile.Close()
+
+	return ParseFatbinFromELFFile(elfFile)
+}
+
+// ParseFatbinFromPath parses the fatbin sections of the given ELF file and returns the information found in it
+func ParseFatbinFromELFFile(elfFile *elf.File) (*Fatbin, error) {
+	fatbin := &Fatbin{
+		Kernels: make(map[CubinKernelKey]*CubinKernel),
+	}
+
+	for _, sect := range elfFile.Sections {
+		// CUDA embeds the fatbin data in sections named .nv_fatbin or __nv_relfatbin
+		if sect.Name != ".nv_fatbin" && sect.Name != "__nv_relfatbin" {
+			continue
+		}
+
+		data, err := sect.Data()
+		if err != nil {
+			return nil, fmt.Errorf("failed to read section %s: %w", sect.Name, err)
+		}
+
+		// The fatbin format will have a header, then a sequence of data headers + payloads.
+		// After the data corresponding to the first header, we might have more headers so
+		// we need to loop until we reach the end of the section data
+		buffer := bytes.NewReader(data)
+		for buffer.Len() > 0 {
+			var fatbinHeader fatbinHeader
+			err = binary.Read(buffer, binary.LittleEndian, &fatbinHeader)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse fatbin header: %w", err)
+			}
+
+			// Check the header is valid
+			if fatbinHeader.Magic != fatbinMagic {
+				return nil, fmt.Errorf("invalid fatbin header, magic number %x does not match expected %x", fatbinHeader.Magic, fatbinMagic)
+			}
+			if fatbinHeader.Version != fatbinHeaderVersion {
+				return nil, fmt.Errorf("invalid fatbin header, version %d does not match expected %d", fatbinHeader.Version, fatbinHeaderVersion)
+			}
+
+			// We need to read only up to the size given to us by the header, not to the end of the section.
+			readStart, _ := buffer.Seek(0, io.SeekCurrent)
+			for currOffset, _ := buffer.Seek(0, io.SeekCurrent); buffer.Len() > 0 && currOffset-readStart < int64(fatbinHeader.FatSize); currOffset, _ = buffer.Seek(0, io.SeekCurrent) {
+				// Each data section starts with a data header, read it
+				var fatbinData fatbinData
+				err = binary.Read(buffer, binary.LittleEndian, &fatbinData)
+				if err != nil {
+					return nil, fmt.Errorf("failed to parse fatbin data: %w", err)
+				}
+
+				// Check that we have a valid header
+				dataKind := fatbinDataKind(fatbinData.Kind)
+				if dataKind < fatbinDataMinKind || dataKind > fatbinDataMaxKind {
+					return nil, fmt.Errorf("invalid fatbin data, kind %d is not in the expected range [%d, %d]", dataKind, fatbinDataMinKind, fatbinDataMaxKind)
+				}
+				if fatbinData.Version != fatbinDataVersion {
+					return nil, fmt.Errorf("invalid fatbin data, version %d does not match expected %d", fatbinData.Version, fatbinDataVersion)
+				}
```

Comment:
consider defining Validate function for both fatbinData and fatbinHeader structs and move the relevant logic into those functions